src/launcher/java/org/truffleruby/launcher/options/CommandLineOptions.java

```diff
@@ -32,70 +32,57 @@
 import java.util.List;
 import java.util.Map;
+import java.util.stream.Collectors;
 
 public class CommandLineOptions {
 
-    private Map<String, String> options = new HashMap<>();
+    private final Map<String, CharSequence> options = new HashMap<>();
     private String[] arguments = new String[]{};
     private final List<String> unknownArguments = new ArrayList<>(0);
 
-    // TODO (pitr-ch 26-Jul-2017): move as much as possible to options
-
-    private StringBuffer inlineScript = new StringBuffer();
-    private boolean hasInlineScript;
-    private boolean usePathScript;
-    private String scriptFileName;
-    private boolean showVersion;
-    private boolean showCopyright;
-    private boolean shouldRunInterpreter = true;
-    private boolean shouldPrintUsage;
-    private boolean shouldCheckSyntax;
-    private String inPlaceBackupExtension;
-    private boolean hasScriptToRun; // -e or a file
-    private boolean forceStdin;
-    private boolean shouldPrintShortUsage;
-
     // TODO (pitr-ch 26-Jul-2017): Move to Options when implementing -s option
     // Currently not used
     private Map<String, String> optionGlobals = new HashMap<>();
 
-    public String getDisplayedFileName() {
-        if (isInlineScript()) {
-            if (getScriptFileName() != null) {
-                return getScriptFileName();
-            } else {
-                return "-e";
-            }
-        } else if (shouldUsePathScript()) {
-            return "-S";
-        } else if (isForceStdin() || getScriptFileName() == null) {
-            return "-";
-        } else {
-            return getScriptFileName();
-        }
+    Map<String, CharSequence> getOptionsInternal() {
+        return options;
     }
 
     public Map<String, String> getOptions() {
-        return options;
+        return options.entrySet().stream().
+                collect(Collectors.toMap(Map.Entry::getKey, (e) -> e.getValue().toString()));
     }
 
     public <T> void setOption(OptionDescription<T> key, T value) {
         setOptionRaw(key, key.toString(value));
     }
 
-    public void setOptionRaw(OptionDescription<?> key, String value) {
-        options.put(key.getName(), value);
-    }
-
     public <T> T getOption(OptionDescription<T> key) {
         return key.checkValue(getOptionRaw(key));
     }
 
-    public void appendOptionValue(StringArrayOptionDescription key, String newValue) {
-        setOptionRaw(key, key.append(getOptionRaw(key), newValue));
+    public <T> void appendOptionValue(AppendableOptionDescription<T> key, String newValue) {
+        StringBuilder builder;
+        CharSequence current = getOptionRaw(key);
+
+        if (current == null) {
+            builder = new StringBuilder();
+            setOptionRaw(key, builder);
+        } else if (current instanceof StringBuilder) {
+            builder = (StringBuilder) current;
+        } else {
+            builder = new StringBuilder(current);
+            setOptionRaw(key, builder);
+        }
```

Comment:
It makes this weird CharSequence signature and needs copying at the high-level, I think it's highly premature, few options are appended and very few times, isn't it?